### PR TITLE
Fix nesting of cleaned directories.

### PR DESCRIPTION
**Pull request by ClarenceGardner**
https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B0000005RWDbIAO/view

### CODE REVIEW

File: o4/o4.py [hunk 1]
```diff
@@ -1069,8 +1069,8 @@ def move_except(from_dir, to_dir, but_not):
             sys.exit(f'*** ERROR: Cannot resume cleaning; {source} does not exist.')
     else:
         os.makedirs(f'{source}/.o4', exist_ok=True)
-        move_except(f'{target}/.o4', f'{source}/.o4', but_not='cleaning')
-        move_except(target, source, but_not='.o4')
+        move_except(f'{target}/.o4', f'{source}/.o4', but_not=('cleaning', 'cleaned'))
+        move_except(target, source, but_not=['.o4'])
```
Comment (philipbergen) [inline o4/o4.py]: Lists or tuples, buddy? :-D

File: o4/o4.py [hunk 2]
```diff
@@ -1092,13 +1092,43 @@ def move_except(from_dir, to_dir, but_not):
     check_call(cmd)
     if not discard:
         savedir = source.replace('cleaning', time.strftime('cleaned'))  # @%Y-%m-%d,%H:%M'))
+        if os.path.exists(savedir):
+            shutil.rmtree(savedir)
```
Comment (philipbergen) [inline o4/o4.py]: This is dangerous, they will lose data. I'd rather we just moved it to `.o4/1-cleaned` or something like it

File: o4/o4.py [hunk 3]
```diff
@@ -1092,13 +1092,43 @@ def move_except(from_dir, to_dir, but_not):
...
         assert source.endswith('cleaning')
         shutil.rmtree(source)
 
 
+def rm_empty_dirs(root):
+    '''
+    Given a directory tree, remove all directories that have only
+    directories beneath them. Return the number of non-directory files
+    remaining.
+    '''
```
Comment (philipbergen) [inline o4/o4.py]: `find . -type -d -exec rmdir -p '{}' ';'` ?

File: o4/version.py [hunk 4]
```diff
@@ -3,13 +3,13 @@
 
 import datetime
 
-VERSION = (1, 2, 144)
-VERSION_STR = '1.2.144'
+VERSION = (1, 1, 0)
+VERSION_STR = '1.1.0'
```
Comment (philipbergen) [inline o4/version.py]: How'd you make it go backwards in versions? :D

File: o4/o4.py (hunk 2 above)
Comment (philipbergen) [reply o4/o4.py]: Am I the only one that has a panic attack before testing anything with rmtree/`rm -fr` in it?

File: o4/o4.py (hunk 1 above)
Comment (ClarenceGardner) [reply o4/o4.py]: Sequences

File: o4/o4.py (hunk 2 above)
Comment (ClarenceGardner) [reply o4/o4.py]: I test it with a print first

Comment (philipbergen) [reply o4/o4.py]: And still I basically need new underwear after I tried it for realz

File: o4/o4.py (hunk 3 above)
Comment (ClarenceGardner) [reply o4/o4.py]: Didn't know about -depth.
Man page on mac is quite useful:
  -d same -as -depth
  -depth same as -d
It seems pretty weird to combine that with rmdir -p; I suppose find doesn't get confused.

This works, and returns the number of files, so I'm inclined to leave it unless you have a good reason not to.

Comment (philipbergen) [reply o4/o4.py]: Leave it. I just wanted to share how I would have done it. Not saying that’s how you would have done it. I was just too lazy to do your real solution. 

File count is great. Might even add a fstat and compute total size?

File: o4/o4.py [hunk 5]
```diff
@@ -1159,6 +1159,13 @@ def move_except(from_dir, to_dir, but_not):
 
     target = os.getcwd()
     source = f'{target}/.o4/cleaning'
+    cleaned = f'{target}/.o4/cleaned'
+    if rm_empty_dirs(cleaned):
+        print(f'*** ERROR: Unhandled files still exist from a previous clean run.')
+        print(f'           Please delete (if unwanted) or move files from {cleaned}')
+        print(f'           back to its place in {target}.')
+        sys.exit(1)
+
```
Comment (ClarenceGardner) [inline o4/o4.py]: I prefer this to renaming an existing "cleaned" -- possibly more than once. Really, if the clean left detritus, it should be taken care of by the user.

Comment (philipbergen) [reply o4/o4.py]: Agreed. This is very elegant. 
Is there a -f option?

File: o4/o4.py (hunk 3 above)
Comment (ClarenceGardner) [reply o4/o4.py]: I don't see how that would add value. The person needs to address the files regardless.

File: o4/o4.py (hunk 5 above)
Comment (ClarenceGardner) [reply o4/o4.py]: You really want one? I'm thinking that could be too easy for people to do, just to get things going. Kind of "-f fatigue".

Comment (philipbergen) [reply o4/o4.py]: Better not then. We can always add it if we change our minds.

File: o4/o4.py [hunk 6]
```diff
@@ -1187,13 +1194,46 @@ def move_except(from_dir, to_dir, but_not):
     check_call(cmd)
     if not discard:
         savedir = source.replace('cleaning', time.strftime('cleaned'))  # @%Y-%m-%d,%H:%M'))
+        if os.path.exists(savedir):
```
Comment (philipbergen) [inline o4/o4.py]: This can't happen now, right?

File: o4/o4.py [hunk 7]
```diff
@@ -1187,13 +1194,46 @@ def move_except(from_dir, to_dir, but_not):
...
+            shutil.rmtree(savedir)
         shutil.move(source, savedir)
-        err_print(f'*** INFO: Directory is clean @{changelist}; detritus is in {savedir}')
+        if os.path.exists(f'{savedir}/.o4'):
```
Comment (philipbergen) [inline o4/o4.py]: It just has fstat files in it with the above change, right?

File: o4/o4.py (hunk 3 above)
Comment (philipbergen) [reply o4/o4.py]: Yes, with the above error, now they do. Like that a lot.

Comment (philipbergen) [reply o4/o4.py]: Though of course by that logic, the file count likely adds no value either. :D

File: o4/o4.py (hunk 6 above)
Comment (ClarenceGardner) [reply o4/o4.py]: I think it can; at least, I think I was seeing it. I was very confused.

File: o4/o4.py [hunk 8]
```diff
@@ -1154,18 +1154,25 @@ def o4_clean(changelist, quick=False, resume=False, discard=False):
...
+        sys.exit(1)
+
     if resume:
         if not os.path.exists(source):
             sys.exit(f'*** ERROR: Cannot resume cleaning; {source} does not exist.')
     else:
         os.makedirs(f'{source}/.o4', exist_ok=True)
-        move_except(f'{target}/.o4', f'{source}/.o4', but_not='cleaning')
-        move_except(target, source, but_not='.o4')
+        move_except(f'{target}/.o4', f'{source}/.o4', but_not=('cleaning', 'cleaned'))
```
Comment (philipbergen) [inline o4/o4.py]: rm

**Pull request merged**